kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,994 @@
+        // some fields are missing, but they might be nullable, need to insert them into the reorder_indices
+        for (requested_position, field) in requested_schema.fields().enumerate() {
+            if !found_fields.contains(field.name()) {
+                if field.nullable {
+                    debug!("Inserting missing and nullable field: {}", field.name());
+                    reorder_indices.push(ReorderIndex::Null {
+                        index: requested_position,
+                        field: Arc::new(field.try_into()?),
+                    });
+                } else {
+                    return Err(Error::Generic(format!(
+                        "Requested field not found in parquet schema, and field is not nullable: {}",
+                        field.name()
+                    )));
+                }
+            }
+        }
+    }
+    Ok((
+        parquet_offset + fields.len() - start_parquet_offset,
+        reorder_indices,
+    ))
+}
+
+/// Get the indices in `parquet_schema` of the specified columns in `requested_schema`. This returns
+/// a tuples of (mask_indices: Vec<parquet_schema_index>, reorder_indices:
+/// Vec<requested_index>). `mask_indices` is used for generating the mask for reading from the
 /// parquet file, and simply contains an entry for each index we wish to select from the parquet
-/// file set to the index of the requested column in the parquet. `reorder_indicies` is used for
-/// re-ordering and will be the same size as `requested_schema`. Each index in `reorder_indicies`
-/// represents a column that will be in the read parquet data at that index. The value stored in
-/// `reorder_indicies` is the position that the column should appear in the final output. For
-/// example, if `reorder_indicies` is `[2,0,1]`, then the re-ordering code should take the third
-/// column in the raw-read parquet data, and move it to the first column in the final output, the
-/// first column to the second, and the second to the third.
+/// file set to the index of the requested column in the parquet. `reorder_indices` is used for
+/// re-ordering. See the documentation for [`ReorderIndex`] to understand what each element in the
+/// returned array means
 pub(crate) fn get_requested_indices(
     requested_schema: &SchemaRef,
     parquet_schema: &ArrowSchemaRef,
-) -> DeltaResult<(Vec<usize>, Vec<usize>)> {
-    let (mask_indicies, reorder_indicies): (Vec<usize>, Vec<usize>) = parquet_schema
-        .fields()
-        .iter()
-        .enumerate()
-        .filter_map(|(parquet_index, field)| {
-            requested_schema
-                .index_of(field.name())
-                .map(|index| (parquet_index, index))
-        })
-        .unzip();
-    require!(
-        mask_indicies.len() == requested_schema.fields.len(),
-        Error::generic("Didn't find all requested columns in parquet schema")
-    );
-    Ok((mask_indicies, reorder_indicies))
+) -> DeltaResult<(Vec<usize>, Vec<ReorderIndex>)> {
```

Comment:
Hmmm, I don't know if rolling this isn't a struct for internal use would make sense. Especially since you are likely going to just assign it in something like `let (mask, reordered_indexes) = get_requested_indices(...);`

Reply:
Makes sense, though you can always de-structure structs too:

```rust
let SchemaIndex { mask, reordered_indices } = get_requested_indices(...);
```

Reply:
Yeah, given this purely internal for now, I think leaving it as a tuple with a clear doc comment is okay, but lmk if you feel strongly about it :)